### ISSUE TITLE
Clicking a button with mouse should not steal focus from other fields

### ISSUE THREAD

**Issue by vaadin-bot** [vaadin/framework#1113]
**Originally by @jouni**

___
Clicking a button with the mouse should preserve the focus in the field that it is currently, if the button and the field are inside the same window.

If no field is focused, the button shouldn't receive focus in that case either. Instead, the underlying window should receive the focus.

Use case: user is typing a long text in a text field, when he has to press some button for a quick task. If the focus moves to the button, the caret position is lost in the long text, causing the user to lose his context.

Tabbing to the button using the TAB key should focus the button, and apply the proper focus style.

This is most probably very hard to control efficiently, and would require major event propagation designing. But I'm still documenting this as defect, since this is how traditional desktop buttons behave. The browser environment is a different beast, but we are in the business of emulating the desktop.

This will also affect text field value changes, if we won't explicitly cause such events when a button is pressed (, which happen automatically currently because text fields are blurred.

An actual working implementation can be found in SproutCore.

There are also multiple other places where a click steals the focus unnecessarily:
 - expanding/collapsing a tree node
 - clicking a checkbox
 - clicking a link (usually navigates somewhere, so it's not that dramatic)
 - scrolling a panel or any other scrollable element
 - collapsing table columns
 - sorting a table column
 - dragging a splitter
 - scrolling tabs
 - switching tabs
 - activating a MenuBar command
 - dragging a sub-window

So basically anything that involves a mousedown operation.

A jQuery POC, covering a button and checkbox cases, can be found here: http://koivuviita.com/upload/focus-handling/
